test/test_strerror.cpp

```diff
@@ -54,3 +61,38 @@ TEST(test_strerror, get_error) {
     "' did not cause the expected error message.";
 #endif
 }
+
+/*
+   Define the blueprint of a mock identified by `strerror_r_proto`
+   strerror_r signature:
+   int strerror_r(int errnum, char *buf, size_t buflen);
+*/
+mmk_mock_define(strerror_r_mock, int, char *, size_t);
+
+/* Mocking test example */
+TEST(test_strerror, test_mock) {
+  /* Mock the strerror_r function in the current module using
+     the `strerror_r_mock` blueprint. */
+  // mmk_mock("strerror_r@rcutils", strerror_r_mock);
+  // mmk_mock("__xpg_strerror_r@self", strerror_r_mock);
+  // mmk_mock("strerror_r@self", strerror_r_mock);
+  mmk_mock("__xpg_strerror_r@lib:rcutils", strerror_r_mock);
+
+  /* Tell the mock to return NULL and set errno to ENOMEM
+     whatever the given parameter is. */
+  mmk_when(
+    strerror_r(mmk_any(int), mmk_any(char *), mmk_any(size_t) ),
+    .then_return = mmk_val(int, EINVAL));
+
+  // Now normal usage of the function returning unexpected EINVAL
+  // error for the internal strerror_r
+  // This works only for POSIX
+
+  // Set the error "No such file or directory"
+  errno = 2;
+  char error_string[1024];
+  rcutils_strerror(error_string, sizeof(error_string));
+  ASSERT_STREQ(error_string, "Failed to get error");
```

Comment:
I'm sure this is a case-by-case matter, but what risks are there for maintainers adding calls to the mocked function that sits between the unit test and the targeted invocation? It seems like unit tests that call high-level functions, but mock low-level functions would be pretty brittle, and you would need to mock functions whose number of uses in the code under test is unlikely to change. `rcutils_strerror` is a good example where the number of uses of `strerror_r` are unlikely to change. Whereas mocking something like `rcutils_strndup` during complex code like `rclcpp::init` would result in behaviors that are difficult to assert.

Reply:
I don't see a fundamental risk, besides the maintenance burden that comes from white-box testing and the coupling between test and implementation it entails. It's important to bear in mind that mocking is not a tool for fault injection in a running system, but a way to unit test the interactions between the entity under test and its direct dependencies. In an `rclcpp::init` test, one would mock `rcutils_strndup` along with all other direct dependencies to ensure correct behavior.